### DISCUSSION TITLE
[Rust] Easy way to get random number.

### DISCUSSION THREAD

**Discussion by microgift** [community/community#49489] [Apps, API and Webhooks]
### Select Topic Area

Question

### Body

While I was making a Solana program I had to get a random number in rust.
I tried to look switchboard examples but it's not that easy.
Is there any easy way to get it?
I hashed the timestamp and get the last 3 bytes.

**Comment by runesatsdev** [comment]
Generating random number is not trivial in any language or blockchain platform as it requires a source of entropy to generate unpredictable results. In addition to using timestamp like you suggested, you can also consider the following alternatives depending on your specific use case:

Use a trusted oracle - You can use an oracle service like Chainlink VRF (Randomness Function) to securely generate a provably fair random number by generating it off-chain and pushing it onto the Solana blockchain.

Use a combination of user input and hashing- In some cases you might be able to use user-input (which cannot be controlled by the smart contract) to act as a random seed for a hash function such as SHA256.

Use the Solana randomization functions - Solana has built in random functions that are better suited for generating randomness. In rust for example to generate a random integer between 0 and 999 inclusive you can use:
////////////////////////////////////////////////////////////
 #[derive(AnchorSerialize, AnchorDeserialize)]
 struct MyProgram {
     // accounts, etc.
 }

 #[program]
 pub async fn my_method(ctx: Context<MyProgram>) -> ProgramResult {
     let random_number = loader::pubkey().to_bytes()[..4].try_into()?;
     let mut rng = ChaChaRng::from_seed(random_number);
     let rand_integer: u32 = rng.gen_range(0, 1000);
     // ..rest of code here
 }
/////////////////////////////////////////////////////////


It's important to note that generating a truly random output on a deterministic blockchain like Solana is a hard problem and there are different solutions with different tradeoffs.

Best of lucky.

**Comment by ParsaAminpour** [comment]
I came across an on-chain randomness protocol called `Feed`, and it was super useful
You can check it out here: https://feedprotocol.xyz/
